### PR TITLE
Maintenance/form builder upgrades

### PR DESCRIPTION
**Pull request by jordanlesich**

### CODE REVIEW

**Comment by jordanlesich** [comment]
- Validation Overhaul
- Modifiers Overhaul
- MultiTX form
- React-Hook-Form nested data rebuild for FormBuilder

File: src/data/fields.js [hunk 1]
```diff
@@ -424,11 +422,7 @@ export const FIELD = {
       signal: 'not checked',
     },
     title: 'Create a token proposal',
-    description: {
-      type: 'formCondition',
-      token: 'This creates a WhiteList Token TX',
-      signal: 'This creates a signal proposal',
-    },
+    description: 'ssdfsd',
```
Comment (skuhlmann) [inline src/data/fields.js:425]: Should this be here?

File: src/formBuilder/checkGate.jsx [hunk 2]
```diff
@@ -1,10 +1,17 @@
 import React, { useState } from 'react';
-import { useFormConditions } from '../utils/formBuilder';
+import { createRegisterOptions, useFormConditions } from '../utils/formBuilder';
```
Comment (skuhlmann) [inline src/formBuilder/checkGate.jsx:2]: import order

File: src/formBuilder/fieldWrapper.jsx [hunk 3]
```diff
@@ -15,18 +16,26 @@ const FieldWrapper = ({
...
   w,
 }) => {
   const width = useMemo(() => {
     if (w) return w;
     // if (layout === 'singleRow') return '100%';
     // return ['100%', null, '48%'];
   }, [w, layout]);
+
+  // const errorMsgs = useMemo(() => {
+  //   if (!errors[name]) return;
```
Comment (skuhlmann) [inline src/formBuilder/fieldWrapper.jsx:33]: tidy me up

File: src/formBuilder/abiInput.jsx [hunk 4]
```diff
@@ -38,29 +41,28 @@ const AbiInput = props => {
       }
     };
     if (targetContract && validate.address(targetContract)) {
+      if (targetContract === contractRef.current) return;
       getABI();
-    } else {
-      setIsDisabled(true);
+      contractRef.current = targetContract;
     }
   }, [targetContract]);
```
Comment (skuhlmann) [inline src/formBuilder/abiInput.jsx:48]: More of a question here i've been struggling with:

Should we add daochain in the dependency array? stuff from params should never trigger a re-render, but i think it's technically a dependency.

File: src/formBuilder/multiForm.jsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
+import React, { useEffect, useState } from 'react';
+import { useForm } from 'react-hook-form';
+
+import { isLastItem } from '../utils/general';
+import { serializeTXs } from '../utils/formBuilder';
+import FormSection from './formSection';
```
Comment (skuhlmann) [inline src/formBuilder/multiForm.jsx:6]: import order

File: src/formBuilder/multiInput.jsx [hunk 6]
```diff
@@ -1,25 +1,44 @@
...
+  ];
+};
+
+const MultiInput = props => {
+  const { name } = props;
+  const [inputs, setInputs] = useState(null);
 
+  // const values = parentForm?.watch();
+  useEffect(() => {
+    console.log('COMPONENT MOUNT');
```
Comment (skuhlmann) [inline src/formBuilder/multiInput.jsx:33]: tidy me up

File: src/formBuilder/paymentInput.jsx [hunk 7]
```diff
@@ -1,33 +1,43 @@
...
 import ModButton from './modButton';
+
+import { getContractBalance } from '../utils/tokenValue';
+import { validate } from '../utils/validation';
 import { handleDecimals } from '../utils/general';
+import { spreadOptions } from '../utils/formBuilder';
 
-const getMaxBalance = (tokenData, tokenAddress) => {
-  //  Uses token select data structure
+// const getMaxBalance = (tokenData, tokenAddress) => {
```
Comment (skuhlmann) [inline src/formBuilder/paymentInput.jsx:12]: clean up

File: src/formBuilder/paymentInput.jsx [hunk 8]
```diff
@@ -57,19 +67,35 @@ const PaymentInput = props => {
...
+      if (token?.balance && token.decimals) {
+        setToken(token);
+      }
     }
   }, [daoTokens, paymentToken]);
 
   const setMax = () => {
-    setValue('paymentRequested', balance);
+    if (!token?.balance) return;
+    setValue('paymentRequested', token.balance / 10 ** token.decimals);
```
Comment (skuhlmann) [inline src/formBuilder/paymentInput.jsx:80]: ah great fix. getMaxBalance was a bad issue

File: src/formBuilder/tributeInput.jsx [hunk 9]
```diff
@@ -134,12 +136,28 @@ const TributeInput = props => {
...
+        getContractBalance(val, decimals) > Number(allowance)
+          ? `Amount entered exceeds token allowance.`
+          : true,
+      hasBalance: val =>
+        getContractBalance(val, decimals) > Number(balance)
+          ? `Amount entered exceeds wallet balance.`
+          : true,
+    },
+  });
+  console.log(`options`, options);
```
Comment (skuhlmann) [inline src/formBuilder/tributeInput.jsx:153]: clean

File: src/formBuilder/txFormSection.jsx [hunk 10]
```diff
@@ -0,0 +1,70 @@
+import React, { useEffect, useState } from 'react';
+import { Button, Flex, Icon } from '@chakra-ui/react';
+import { BiMinus, BiPlus } from 'react-icons/bi';
+
+import TextBox from '../components/TextBox';
+
+import { IsJsonString } from '../utils/general';
+import FormSection from './formSection';
```
Comment (skuhlmann) [inline src/formBuilder/txFormSection.jsx:8]: import order

File: src/forms/lootGrab.jsx [hunk 11]
```diff
@@ -1,3 +1,4 @@
+//  Purge
```
Comment (skuhlmann) [inline src/forms/lootGrab.jsx:1]: PURGE

File: src/forms/memberProposal.jsx [hunk 12]
```diff
@@ -1,3 +1,5 @@
+//  Purge
```
Comment (skuhlmann) [inline src/forms/memberProposal.jsx:1]: AGAIN

File: src/forms/tradeProposal.jsx [hunk 13]
```diff
@@ -1,3 +1,5 @@
+// Purge
```
Comment (skuhlmann) [inline src/forms/tradeProposal.jsx:1]: PURGE

File: src/utils/tokenValue.js [hunk 14]
```diff
@@ -2,6 +2,7 @@ import { ethers } from 'ethers';
 import { TokenService } from '../services/tokenService';
 import { MolochService } from '../services/molochService';
```
Comment (skuhlmann) [inline src/utils/tokenValue.js:3]: Can we refactor to contract service? 

Not necessary immediately.

File: src/utils/txHelpers.js [hunk 15]
```diff
@@ -86,6 +87,36 @@ const buildJSONdetails = (data, fields) => {
...
+
+const collapseToCallData = values =>
+  values.TX.map(tx => ({
+    to: tx.targetContract,
+    data: safeEncodeHexFunction(JSON.parse(tx.abiInput), tx.abiArgs),
+    value: tx.minionValue || '0',
+    operation: tx.operation || '0',
+  }));
+
+// console.log(
```
Comment (skuhlmann) [inline src/utils/txHelpers.js:113]: can clean up

File: src/utils/validation.js [hunk 16]
```diff
@@ -213,3 +159,13 @@ export const customValidations = {
     return false;
   },
 };
+
+export const buildRHFvalFn = valString => {
```
Comment (skuhlmann) [inline src/utils/validation.js:163]: function name is a little confusing!

**Review by skuhlmann — CHANGES_REQUESTED** [review]

File: src/data/fields.js (hunk 1 above)
Comment (jordanlesich) [reply src/data/fields.js:425]: lol

File: src/formBuilder/abiInput.jsx (hunk 4 above)
Comment (jordanlesich) [reply src/formBuilder/abiInput.jsx:48]: Yep. Makes sense.

File: src/utils/tokenValue.js (hunk 14 above)
Comment (jordanlesich) [reply src/utils/tokenValue.js:3]: Oh, right. Yep.

File: src/utils/validation.js (hunk 16 above)
Comment (jordanlesich) [reply src/utils/validation.js:163]: Right.

File: src/formBuilder/StepperForm.jsx [hunk 17]
```diff
@@ -34,6 +34,7 @@ const StepperForm = props => {
     metaFields,
     updateModalUI,
   } = props;
+  const [localSteps, setLocalSteps] = useState(steps);
```
Comment (skuhlmann) [inline src/formBuilder/StepperForm.jsx:37]: src/formBuilder/StepperForm.jsx
  Line 37:22:  'setLocalSteps' is assigned a value but never used  no-unused-vars

File: src/utils/tokenValue.js (hunk 14 above)
Comment (jordanlesich) [reply src/utils/tokenValue.js:3]: I'm going to hold out on this for a bit. Would require to me to test and make sure all things calling these utils are getting the right shape of data. But I think I need to go on a larger purge spree, so this should be prime pickens.

File: src/formBuilder/StepperForm.jsx (hunk 17 above)
Comment (jordanlesich) [reply src/formBuilder/StepperForm.jsx:37]: good catch.

**Pull request closed without merging**